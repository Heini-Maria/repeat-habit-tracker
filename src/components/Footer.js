import { StyleSheet, View, Text } from "react-native";

import Icon from 'react-native-vector-icons/FontAwesome';

export default function Footer({ navigation }) {
    return(
        <View style={styles.footer}>
            <Icon 
            name='home'
            size= {30}
            color= "#463C33"
            onPress= {()  => navigation.navigate('Home')}
            />
        </View>
    )
}

const styles = StyleSheet.create({
    footer: {
        flex: 2,
        width: '100%',
        alignItems: 'center',
        justifyContent: 'center',
        backgroundColor: '#9CBFDF', 
        zIndex: 5,
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 2,
        },
        shadowOpacity: 0.25,
        shadowRadius: 3.84,

        elevation: 5,   
    },
    text: {
        fontSize: 18,
        fontWeight: '900'
    }
})