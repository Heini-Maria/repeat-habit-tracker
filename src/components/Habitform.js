import React, { useEffect, useState, useRef } from "react"
import { StyleSheet, View, Text, TextInput, TouchableOpacity, FlatList, Switch, Image} from 'react-native';
import Modal from 'react-native-modal';
import SelectDropdown from "react-native-select-dropdown";
import FontAwesome from 'react-native-vector-icons/FontAwesome';


export default function Habitform ({ GlobalState, isVisible, setIsVisible }) {
    const {habitList, setHabitList, habit, setHabit, goal, setGoal, frequency, setFrequency,} = GlobalState;
    const goals = ['1', '2', '3', '4', '5', '6', '7'];
    const per = ['day', 'week', 'month']
    const dropdownRef = useRef();
    const [habitError, setHabitError] = useState('');
   
    const handleChange = (input) => {
        setHabit(input);
        setHabitError('');
    }

    const handleSaveHabit = () => {
        const index = habitList.length + 1;
        const regex = /^[a-zA-z]+$/;
        let isValid = regex.test(habit)
        console.warn(isValid);
        
        if(isValid && habit !== '' && goal !== '' && frequency !== '') {
        setHabitList(prevState => [...prevState, {id: index, habit: habit, times: 0, goal: goal, frequency: frequency, completed: false}]);
        setHabit('');
        setGoal('');
        setFrequency('')
        dropdownRef.current.reset();
        setIsVisible(!isVisible);
        setHabitError('');
        console.log(habitList);
    } else if (habit === '' && goal === ''){
        setIsVisible(!isVisible);
    } else if (habit !== '' && !isValid) {
        setHabitError('Habit should contain only letters!');
    } else {
        setHabitError('Please fill all fields below');
    }
    }

    return (
           <View style={styles.inputs}  >
                <Modal isVisible={isVisible}
                      style={styles.modal}  >
                <Text style={styles.error}>{habitError}</Text>
                <TextInput
                    style={styles.input}
                    onChangeText={handleChange}
                    value={habit}
                    keyboardType= 'email-address'
                    placeholder= "habit to add..."
                    maxLength={20}
                />
                <SelectDropdown
                    data= {goals}
                    ref= {dropdownRef}
                    defaultButtonText= {'select # of times'}
                    buttonTextStyle ={styles.dropdownText}
                    buttonStyle={styles.dropdown}
                    onSelect={(selectedItem, index) => {
                        console.log(selectedItem, index);
                        setGoal(selectedItem);
                    }}
                    renderDropdownIcon={isOpened => {
                        return <FontAwesome name={isOpened ? 'chevron-up' : 'chevron-down'} color={'#463C33'} size={12} />;
                      }}
                />
                <SelectDropdown
                    data= {per}
                    ref= {dropdownRef}
                    defaultButtonText= {'select timeframe'}
                    buttonTextStyle ={styles.dropdownText}
                    buttonStyle={styles.dropdown}
                    onSelect={(selectedItem, index) => {
                        console.log(selectedItem, index);
                        setFrequency(selectedItem);
                    }}
                    renderDropdownIcon={isOpened => {
                        return <FontAwesome name={isOpened ? 'chevron-up' : 'chevron-down'} color={'#463C33'} size={12} />;
                      }}
                />
                <TouchableOpacity
                    style={styles.button}
                    onPress={handleSaveHabit}
                >
                    <FontAwesome name= 'plus' color={'#F3F3F4'} size={20} />
                </TouchableOpacity>
                </Modal>
        </View>
    )
}
const styles = StyleSheet.create({
    inputs: {
        flex: 3,
        width: '100%',
        backgroundColor: '#F3F3F4',
        padding: 15,
        flexWrap: 'wrap',

    },
    text: {
        fontFamily: 'Inter',
        marginBottom: 20,
        fontSize: 20,
        textAlign: 'left',
        padding: 10,
        color: '#463C33',
        backgroundColor: 'rgba(156,191,223, 0.9)'
     
    },
    input: {
        backgroundColor: 'white',
        padding: 15,
        fontSize: 16,
        width: '90%',
        borderRadius: 12,
        fontFamily: 'Amaranth',
    },
    dropdown: {
        backgroundColor: 'white',
        width: '45%',
        display: 'inline-block',
        marginTop: 10,
        padding: 15,
        borderRadius: 12,
    },
    dropdownText: {
        color: '#463C33',
        fontSize: 16,
        fontFamily: 'Amaranth',
        textAlign: 'left',
    },
    button: {
        backgroundColor: '#FDA769',
        height: 50,
        width: 50,
        marginRight: 30,
        padding: 15,
        borderRadius: '50%',
        alignItems: 'center',
        alignSelf: 'flex-end'  
    },
    error: {
        color: 'tomato', 
        marginLeft: 10, 
        marginBottom: 5,
        fontFamily: 'Inter',
        fontSize: 18,
    }

})