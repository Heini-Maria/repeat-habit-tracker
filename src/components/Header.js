import React from 'react';
import {
  StyleSheet,
  View,
  Text,
  Image,
} from 'react-native';
import Constants from 'expo-constants';

export default function Header() {
  return (
    <View style={styles.header}>
      <Image style={styles.logo} source={require('../../assets/Images/logo.png')}/>
      <Text style={styles.text}>Repeat</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  header: {
    flex: 3,
    width: '100%',
    alignItems: 'center',
    justifyContent: 'flex-start',
    backgroundColor: '#9CBFDF',
    paddingTop: Constants.StatusBarHeight,
    paddingBottom: 10,
    zIndex: 5,
    flexWrap: 'wrap',
    flexDirection: 'row',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
  },
  text: {
    fontSize: 30,
    fontFamily: 'Shrikhand',
    color: '#463C33',
    marginTop: 10,
    position: 'relative',
    left: '90%',
  },
  logo: {
    marginTop: 100,
    width: 35,
    height: 35,
    marginLeft: 36,
  },
});
