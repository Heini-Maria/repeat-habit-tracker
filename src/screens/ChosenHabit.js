import React, { useEffect, useState } from "react"
import { StyleSheet, View, Text, TextInput, TouchableOpacity, FlatList } from 'react-native';

import Header from '../components/Header';
import Footer from '../components/Footer';

export default function ChosenHabit ({ navigation, GlobalState }) {
    const { chosenHabit } = GlobalState;

    return (
        <View style={styles.screen}>
            <Header />
            <View style={styles.body}>
                <Text>{chosenHabit.habit}</Text>
            </View>
            <Footer navigation={navigation} />
        </View>
    )
}
const styles = StyleSheet.create({
    screen: {
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'center',
        justifyContent: 'center',
    },
    body: {
        flex: 8,
        width: '100%',
        backgroundColor: '#F3F3F4',
        alignItems: 'center',
        justifyContent: 'center',
    }, 
})