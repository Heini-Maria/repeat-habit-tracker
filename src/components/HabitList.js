import React, { useEffect } from 'react';
import {
  StyleSheet,
  View,
  Text,
  TouchableOpacity,
  FlatList,
} from 'react-native';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import AsyncStorage from '@react-native-async-storage/async-storage';
import ChosenHabit from "../screens/ChosenHabit";

export default function HabitList({ navigation, GlobalState }) {
  const { habitList, setHabitList, setChosenHabit } = GlobalState;
  const dateCheck = (habitList) => {
    for (let i = 0; i < habitList.length; i++) {
      const value = habitList[i];
      let created = value.created;
      const rightNow = Date.now();
      const oneDay = created + 1000 * 60 * 60 * 24;
      const oneWeek = created + 1000 * 60 * 60 * 24 * 7;
      const oneMonth = created + 1000 * 60 * 60 * 24 * 30;
      if (value.frequency === 'day' && rightNow > oneDay || value.frequency === 'week' && created > oneWeek || value.frequency === 'month' && created > oneMonth) {
        if (value.completed === false) {
          value.completedCount = 0;
        }
        value.created= Date.now();
        value.times = 0;
        value.completed =false;
        setHabitList([...habitList]);
      }
      AsyncStorage.setItem("habits", JSON.stringify(habitList));
    }
  };

  useEffect(() => {
    dateCheck(habitList);
  }, [dateCheck, habitList]);

  const handleDelete = (item) => {
    const index = habitList.indexOf(item);
    habitList.splice(index, 1);
    const newHabits = [...habitList];
    AsyncStorage.setItem('habits', JSON.stringify(newHabits)).then(() => {
      setHabitList(newHabits);
    });
  };

  const handleCheck = (item) => {
    const goal = parseInt(item.goal, 10);
    const times = parseInt(item.times, 10);
    const complete = parseInt(item.goal, 10) - 1;
    if (times === complete) {
      item.completed = true;
      item.completedCount += 1;
      item.times += 1;
      setHabitList([...habitList]);
    }
    else if (times < goal) {
      item.completed = false;
      item.times += 1;
      setHabitList([...habitList]);
    }
    AsyncStorage.setItem('habits', JSON.stringify(habitList));
  };

  const handleChooseHabit = (item) => {
    setChosenHabit(item);
    navigation.navigate('ChosenHabit');
  };

  const renderItem = ({ item }) => {
    return (
      <View style={styles.habit}>
        <TouchableOpacity
          testId="Habit"
          onPress={() => handleChooseHabit(item)}
        >
          <Text style={styles.habitText}>
            {item.habit}
          </Text>
            {item.completed ? 
              <Text style={styles.completed}>Completed</Text>
              : <Text style ={styles.habitGoal}>{item.times} / {item.goal} per {item.frequency}</Text>}
        </TouchableOpacity>
        <View style={styles.icons}>
          <TouchableOpacity style={styles.check} onPress={() => handleCheck(item)}>
            <FontAwesome name="check" color={'#ABC270'} size={30} />
          </TouchableOpacity>
          <TouchableOpacity onPress={() => handleDelete(item)}>
            <FontAwesome name="trash" color={'#463C33'} size={25}/>
          </TouchableOpacity>
        </View>
      </View>
    );
  };

  return (
    <FlatList
      data={habitList}
      renderItem={renderItem}
      keyExtractor={(item) => item.id}
    />
  );
}

const styles = StyleSheet.create({
  habit: {
    backgroundColor: 'white',
    flexDirection: 'row',
    height: 60,
    padding: 10,
    paddingRight: 30,
    margin: 10,
    justifyContent: 'space-between',
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.23,
    shadowRadius: 2.62,
    elevation: 4,
  },
  habitText: {
    fontFamily: 'Amaranth',
    fontSize: 18,
    color: '#463C33',
  },
  habitGoal: {
    fontFamily: 'Inter',
    fontSize: 16,
    color: '#463C33',
    marginTop: 5,
  },
  completed: {
    fontFamily: 'Inter',
    fontSize: 16,
    color: '#FDA769',
    marginTop: 5,
  },
  icons: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  check: {
    marginRight: 20,
  },
  button: {
    backgroundColor: '#FDA769',
    height: 50,
    width: 50,
    marginRight: 40,
    padding: 15,
    borderRadius: '50%',
    alignItems: 'center',
    alignSelf: 'flex-start',
  },
});
