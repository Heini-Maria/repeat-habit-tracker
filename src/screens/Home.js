import React from "react"
import { StyleSheet, View, TouchableOpacity } from 'react-native';
import FontAwesome from 'react-native-vector-icons/FontAwesome';

import Header from '../components/Header';
import Footer from '../components/Footer';
import HabitList from "../components/HabitList";
import Habitform from "../components/Habitform";

export default function Home ({ navigation, GlobalState }) {
    const {/* habitList, setHabitList, 
        habit, setHabit, 
        chosenHabit, setChosenHabit, 
        goal, setGoal,
        frequency, setFrequency,
        completed, setCompleted, */
        isVisible, setIsVisible
    } = GlobalState;
    
    const renderForm  = () => {
        setIsVisible(!isVisible);
        return isVisible;
    }


    return (
        <View style={styles.screen}>
            <Header />
            <View style={styles.body}>
                <HabitList 
                    GlobalState={GlobalState}
                    navigation={navigation}/>
            </View>
            <Habitform 
                isVisible={isVisible}
                setIsVisible={setIsVisible}
                GlobalState={GlobalState}/>
            <View style={styles.addHabit}>    
                <TouchableOpacity
                    style={styles.button}
                    onPress={renderForm}>
                    <FontAwesome name= 'plus' color={'#F3F3F4'} size={20} />
                </TouchableOpacity>
            </View> 
            <Footer navigation={navigation} />
        </View>
    )
}
const styles = StyleSheet.create({
    screen: {
        flex: 1,
        backgroundColor: 'white',
        alignItems: 'center',
        justifyContent: 'center'
    },
    body: {
        flex: 16,
        width: '100%',
        paddingTop: 30,
        backgroundColor: '#F3F3F4',   
    },
    addHabit: {
        flex: 2,
        display: 'block',
        width: '100%',
        padding: 15,
        backgroundColor: '#F3F3F4',
        alignItems: "flex-end"
    },
    button: {
        backgroundColor: '#FDA769',
        height: 50,
        width: 50,
        marginRight: 40,
        padding: 15,
        borderRadius: '50%',
        alignItems: 'center',
        alignSelf: 'flex-end',
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 2,
        },
        shadowOpacity: 0.25,
        shadowRadius: 3.84,
        elevation: 5,  
    },
})