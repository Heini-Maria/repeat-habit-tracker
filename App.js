import React, { useState, useEffect } from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import registerNNPushToken from 'native-notify';
import * as Font from 'expo-font';
import AppLoading from 'expo-app-loading';
import AsyncStorage from "@react-native-async-storage/async-storage";

import Home from './src/screens/Home';
import ChosenHabit from './src/screens/ChosenHabit';


const Stack = createNativeStackNavigator();
const getFonts = () => Font.loadAsync({
    'Amaranth': require('./assets/Fonts/Amaranth/Amaranth-Regular.ttf'),
    'Anton': require('./assets/Fonts/Anton/Anton-Regular.ttf'),
    'Inter': require('./assets/Fonts/Inter/Inter-Regular.ttf'),
    'Shrikhand': require('./assets/Fonts/Shrikhand/Shrikhand-Regular.ttf')
  });


export default function App() {
 
  //push notifications
  registerNNPushToken(5638, 'RspELd7m7YAUK1aICdo8W4');


  // globalstate management
  const [habitList, setHabitList] = useState('');
  const [habit, setHabit] = useState('');
  const [goal, setGoal] = useState('');
  const [frequency, setFrequency] = useState('');
  const [completed, setCompleted] = useState(false);
  const [isVisible, setIsVisible] = useState(false);
  const [chosenHabit, setChosenHabit] = useState('');
  const [fontsLoaded, setFontsLoaded] = useState(false);


  const getHabitList = () => {
    AsyncStorage.getItem('habits').then(data => {
      if(data !== null) {
        setHabitList(JSON.parse(data))
      }
    }).catch((error) => console.log(error));
  }

  useEffect (() => {
    getHabitList();
  }, []);
  
  const GlobalState = {
    habitList, setHabitList,
    habit, setHabit,
    goal, setGoal,
    frequency, setFrequency,
    completed, setCompleted,
    isVisible, setIsVisible,
    chosenHabit, setChosenHabit,
    fontsLoaded, setFontsLoaded,
  }
  

  //navigation
  if(fontsLoaded) {   

  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen name="Home" options= {{headerShown: false}} >
          {props => <Home {...props} GlobalState = {GlobalState} />}
        </Stack.Screen>
        <Stack.Screen name="ChosenHabit" options= {{headerShown: false}} >
          {props => <ChosenHabit {...props} GlobalState = {GlobalState} />}
        </Stack.Screen>
      </Stack.Navigator>
    </NavigationContainer>
  );
} else {
  return (
  <AppLoading
    startAsync= {getFonts}
    onFinish={()=> setFontsLoaded(true)}
    onError={(err) => console.log(err)}
  />
  
 
  )
}
}

