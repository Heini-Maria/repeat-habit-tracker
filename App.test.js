
import 'react-native';
import React from 'react';
import App from './App';
import  Home  from './src/screens/Home';
import AsyncStorage from "@react-native-async-storage/async-storage";

import ChosenHabit from './src/screens/ChosenHabit';
import {create, renderer, shallow} from 'react-test-renderer';
import Habitform from './src/components/Habitform';
import HabitList from './src/components/HabitList';
import Header from './src/components/Header';
import Footer from './src/components/Footer';

jest.runAllTimers();


describe('test for App', () => {
test('Test Rendering', () => {
    const tree = create(<App />)
    expect(tree).toMatchSnapshot();
})
})
describe('test for Header', () => {
    test('Test Rendering', () => {
        const header = create(<Header/>)
        expect(header).toMatchSnapshot();
    })
})
describe('test for Footer', () => {
    test('Test Rendering', () => {
        const footer = create(<Footer/>)
        expect(footer).toMatchSnapshot();
    })
})

describe('test for Footers Home icon', () => {
    const navigation = {
        navigate: jest.fn()
    }
    const tree = create(<Footer navigation={navigation} />)
    test('navigate to home', () => {
        
        const button = tree.root.findByProps({testID: 'home'}).props;
        button.onPress();
        expect(navigation.navigate).toBeCalledWith('Home');
    })
})
