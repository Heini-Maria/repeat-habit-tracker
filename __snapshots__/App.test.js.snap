// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for App Test Rendering 1`] = `null`;

exports[`test for ChosenHabit Test Rendering 1`] = `<chosenHabit />`;

exports[`test for Footer Test Rendering 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#9CBFDF",
      "elevation": 5,
      "flex": 2,
      "justifyContent": "center",
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.25,
      "shadowRadius": 3.84,
      "width": "100%",
      "zIndex": 5,
    }
  }
>
  <Text />
</View>
`;

exports[`test for Header Test Rendering 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#9CBFDF",
      "elevation": 5,
      "flex": 3,
      "flexDirection": "row",
      "flexWrap": "wrap",
      "justifyContent": "flex-start",
      "paddingBottom": 10,
      "paddingTop": undefined,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.25,
      "shadowRadius": 3.84,
      "width": "100%",
      "zIndex": 5,
    }
  }
>
  <Image
    source={1}
    style={
      Object {
        "height": 35,
        "marginLeft": 36,
        "marginTop": 100,
        "width": 35,
      }
    }
  />
  <Text
    style={
      Object {
        "color": "#463C33",
        "fontFamily": "Shrikhand",
        "fontSize": 30,
        "left": "90%",
        "marginTop": 10,
        "position": "relative",
      }
    }
  >
    Repeat
  </Text>
</View>
`;
